### PR TITLE
[REVIEW] Add support to deep-copy struct columns from struct column-view

### LINKED ISSUE #6246: [FEA] Support filter() operator for struct_view
**Is your feature request related to a problem? Please describe.**
Currently, if one passes a struct column as part of a table through the filter operation, the following error is encountered:

```
ai.rapids.cudf.CudfException: cuDF failure at: /ABC/cudf/cpp/src/column/column.cu:247: struct_view not supported yet
	at ai.rapids.cudf.Table.filter(Native Method)
	at ai.rapids.cudf.Table.filter(Table.java:1536)
	at ai.rapids.cudf.TableTest.testStructColumnFilter(TableTest.java:3044)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
```
This issue is a request for filter() feature for struct columns.
**Describe the solution you'd like**
Provide support for struct_view for filter operator.

**Describe alternatives you've considered**
N/A

### PR DESCRIPTION
**Pull request by mythrocks**
Closes #6246.

Filtering a table that contains struct-columns fails because struct columns cannot yet be deep-copied from
a column-view.
This commit fixes the problem.

Tests have been added to confirm that `cudf::apply_bitmask()` does not fail on tables containing struct columns, or even list-of-struct columns.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by kuhushukla** [comment]
I'll try this patch out with my filter test soon here and update.

**Comment by kuhushukla** [comment]
I tried the following _simple_ filter test on java side and the cpp side assert failures are gone and test comes back clean.

```
  @Test
  void testStructColumnFilter() {
    HostColumnVector.ColumnBuilder.StructType type = new HostColumnVector.ColumnBuilder.StructType(true, 4);
    type.addChild(new HostColumnVector.ColumnBuilder.BasicType(true, 4, DType.INT32));
    type.addChild(new HostColumnVector.ColumnBuilder.BasicType(true, 4, DType.INT64));
    HostColumnVector.ColumnBuilder.StructType expectedType = new HostColumnVector.ColumnBuilder.StructType(true, 2);
    expectedType.addChild(new HostColumnVector.ColumnBuilder.BasicType(true, 2, DType.INT32));
    expectedType.addChild(new HostColumnVector.ColumnBuilder.BasicType(true, 2, DType.INT64));
    List data1 = Arrays.asList(10, 20L);
    List data2 = Arrays.asList(50, 60L);
    List data3 = Arrays.asList(10, 80L);
    List data4 = Arrays.asList(20, 80L);
    HostColumnVector.ColumnBuilder.StructData structData1 = new HostColumnVector.ColumnBuilder.StructData(data1);
    HostColumnVector.ColumnBuilder.StructData structData2 = new HostColumnVector.ColumnBuilder.StructData(data2);
    HostColumnVector.ColumnBuilder.StructData structData3 = new HostColumnVector.ColumnBuilder.StructData(data3);
    HostColumnVector.ColumnBuilder.StructData structData4 = new HostColumnVector.ColumnBuilder.StructData(data4);
    try (ColumnVector mask = ColumnVector.fromBoxedBooleans(true, false, true, false);
         ColumnVector fromStructs = ColumnVector.fromStructs(type, Arrays.asList(structData1, structData2, structData3, structData4));
         Table input = new Table(fromStructs);
         Table filteredTable = input.filter(mask);
         ColumnVector expectedStructs = ColumnVector.fromStructs(expectedType, Arrays.asList(structData1, structData3));
         Table expected = new Table(expectedStructs)) {
      assertStructColumnsAreEqual(expected.getColumn(0), filteredTable.getColumn(0), expectedType);
    }
  }
}
```

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,7 @@
 - PR #6119 Add support for `dayofweek` property in `DateTimeIndex` and `DatetimeProperties`
 - PR #6171 Java and Jni support for Struct columns
 - PR #6125 Add support for `Series.mode` and `DataFrame.mode`
+- PR #6271 Add support to deep-copy struct columns from struct column-view
```
Comment (kuhushukla) [inline CHANGELOG.md:14]: Nit: Has a trailing whitespace

**Review by nvdbaranec — APPROVED** [review]
Looks good.  Is there a column_utilities or column test somewhere that could have some tests added to it? I guess they'd be about the same.   Also :  maybe a Struct-of-Lists test?

**Comment by mythrocks** [comment]
>  Is there a column_utilities or column test somewhere that could have some tests added to it? I guess they'd be about the same. Also : maybe a Struct-of-Lists test?

Added both, the former being in `structs_column_tests.cu`.

**Comment by mythrocks** [comment]
How strange. There's an unrelated compile failure in `span_tests.cu`:
```
[ 97%] Building CUDA object tests/CMakeFiles/SPAN_TEST.dir/utilities_tests/span_tests.cu.o
/var/lib/jenkins/workspace/rapidsai/gpuci/cudf/prb/cudf-gpu-build_2/cpp/tests/utilities_tests/span_tests.cu:19:10: fatal error: tests/utilities/base_fixture.hpp: No such file or directory
 #include <tests/utilities/base_fixture.hpp>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by kkraus14** [comment]
> How strange. There's an unrelated compile failure in `span_tests.cu`:
> 
> ```
> [ 97%] Building CUDA object tests/CMakeFiles/SPAN_TEST.dir/utilities_tests/span_tests.cu.o
> /var/lib/jenkins/workspace/rapidsai/gpuci/cudf/prb/cudf-gpu-build_2/cpp/tests/utilities_tests/span_tests.cu:19:10: fatal error: tests/utilities/base_fixture.hpp: No such file or directory
>  #include <tests/utilities/base_fixture.hpp>
>           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```

Fixed in https://github.com/rapidsai/cudf/pull/6304

File: cpp/src/column/column.cu [hunk 2]
```diff
@@ -244,7 +247,26 @@ struct create_column_from_view {
...
+                   });
+
+    auto num_rows = children.empty() ? 0 : children.front()->size();
+
+    return make_structs_column(num_rows,
+                               std::move(children),
+                               view.null_count(),
+                               cudf::copy_bitmask(view.null_mask(), 0, view.size(), stream, mr),
+                               stream,
+                               mr);
```
Comment (shwina) [inline cpp/src/column/column.cu:269]: Does making a structs column involve any device memory allocation? If not, is the `mr` parameter passed just for consistency?

Comment (shwina) [reply cpp/src/column/column.cu:269]: Gotcha. Thanks!

**Review by shwina — APPROVED** [review]
Works great from the Python side!

**Comment by mythrocks** [comment]
Thanks for the reviews, chaps. I've merged this in, now.

**Pull request merged**